pkg/inlinekeepalive/client.go

```diff
@@ -0,0 +1,119 @@
+type KeepaliveClientStream struct {
+	log     hclog.Logger
+	handler grpc.ClientStream
+}
+
+func (k *KeepaliveClientStream) Header() (metadata.MD, error) {
+	return k.handler.Header()
+}
+
+func (k *KeepaliveClientStream) Trailer() metadata.MD {
+	return k.handler.Trailer()
+}
+
+func (k *KeepaliveClientStream) CloseSend() error {
+	return k.handler.CloseSend()
+}
+
+func (k *KeepaliveClientStream) Context() context.Context {
+	return k.handler.Context()
+}
+
+func (k *KeepaliveClientStream) SendMsg(m interface{}) error {
+	return k.handler.SendMsg(m)
+}
+
+// RecvMsg intercepts keepalive messages and does not pass them
+// along to the handler.
+func (k *KeepaliveClientStream) RecvMsg(m interface{}) error {
+	for {
+		err := k.handler.RecvMsg(m)
+		if err != nil {
+			return err
+		}
+
+		pm, ok := m.(protoreflect.ProtoMessage)
+		if !ok {
+			// Weird, not a protobuf message, but not our keepalive, so continue as normal
+			return nil
+		}
+
+		if !IsInlineKeepalive(k.log, pm) {
+			return nil
+		}
+
+		// It's a keepalive message! Ignore it and recv again
+		continue
+	}
+}
+
+// isServerCompatible determines if a server is able to receive inline keepalives
+// by examining the features it advertises.
+func isServerCompatible(ctx context.Context, cc *grpc.ClientConn) (bool, error) {
+	client := pb.NewWaypointClient(cc)
+
+	versionInfo, err := client.GetVersionInfo(ctx, &emptypb.Empty{})
+	if err != nil {
+		return false, errors.Wrapf(err, "failed getting version info to determine if server is inline-keepalive compatible")
+	}
+
+	for _, feature := range versionInfo.Features {
+		if feature == FeatureName {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// KeepaliveClientStreamInterceptor returns a stream interceptor
+// that sends inline keepalive messages on client streams (if the server
+// is compatible), and intercepts inline keepalives from the server.
+func KeepaliveClientStreamInterceptor() grpc.StreamClientInterceptor {
+	return func(
+		ctx context.Context,
+		desc *grpc.StreamDesc,
+		cc *grpc.ClientConn,
+		method string,
+		streamer grpc.Streamer,
+		opts ...grpc.CallOption,
+	) (grpc.ClientStream, error) {
+		log := hclog.FromContext(ctx).With("method", method)
+
+		ctx = metadata.AppendToOutgoingContext(ctx, HeaderSendKeepalives, "true")
+
+		handler, err := streamer(ctx, desc, cc, method, opts...)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get the client handler when setting up the inlinekeepalive interceptor on method %q", method)
+		}
+
+		serverCompatible, err := isServerCompatible(ctx, cc)
+		if err != nil {
+			log.Warn("Failed to determine if server is capatible with inline keepalives - will not send them.", "err", err)
+		}
+
+		// Only send keepalives if this is a client stream - not allowed otherwise.
+		if desc.ClientStreams && serverCompatible {
+			// Send keepalives for as long as the handler has the connection open
+			go ServeKeepalives(ctx, log, handler)
```

Comment:
We need to check the behavior of grpc when calling SendMsg from 2 different goroutines.

Reply:
Good catch - SendMsg is not safe to call concurrently.

I added mutexes to the interceptor to prevent concurrent use.